app/scripts/lib/crypto/recovery-keys.js

```diff
@@ -0,0 +1,175 @@
+          if (! uid) {
+            throw new Error('Missing uid');
+          }
+
+          if (! recoveryKey) {
+            throw new Error('Missing recoveryKey');
+          }
+
+          const km = sjcl.codec.base32.toBits(recoveryKey);
+          const salt = sjcl.codec.hex.toBits(uid);
+
+          const info1 = sjcl.codec.utf8String.toBits('fxa recovery fingerprint');
+          const info2 = sjcl.codec.utf8String.toBits('fxa recovery encrypt key');
+
+          const recoveryKeyIdRaw = sjcl.misc.hkdf(km, 16 * 8, salt, info1);
+          const recoveryKeyEncRaw = sjcl.misc.hkdf(km, 32 * 8, salt, info2);
+
+          // Convert to hex strings for readability
+          const recoveryKeyId = sjcl.codec.hex.fromBits(recoveryKeyIdRaw);
+          const recoveryEnc = sjcl.codec.hex.fromBits(recoveryKeyEncRaw);
+
+          return {
+            recoveryEnc,
+            recoveryKeyId
+          };
+        });
+    },
+
+    /**
+     * Creates an encrypted bundle for the recovery data provided using
+     * the recovery key id and recovery encryption.
+     *
+     * This encryption is based on the protocol described in
+     *
+     * https://github.com/mozilla/fxa-auth-server/blob/master/docs/recovery_keys.md
+     *
+     * @param {String} recoveryKeyId Recovery key id used in encryption
+     * @param {String} recoveryEnc Recovery encryption used to encrypt bundle
+     * @param {Object} recoveryData Data to encrypt
+     * @param {Object} options
+     * @returns {Promise} A promise that will be fulfilled with the encrypted recoveryData
+     */
+    createRecoveryBundle: function (recoveryKeyId, recoveryEnc, recoveryData, options = {}) {
+      return importFxaCryptoDeriver().then((fxaCryptoDeriver) => {
+        const jose = fxaCryptoDeriver.jose;
+
+        if (! recoveryKeyId) {
+          throw new Error('Missing recoveryKeyId');
+        }
+
+        if (! recoveryEnc) {
+          throw new Error('Missing recoveryEnc');
+        }
+
+        const keyOptions = {
+          alg: 'A256GCM',
+          k: jose.util.base64url.encode(recoveryEnc, 'hex'),
+          kty: 'oct'
+        };
+
+        return jose.JWK.asKey(keyOptions)
+          .then(function (key) {
+            const recipient = {
+              header: {
+                alg: 'dir',
+                enc: 'A256GCM',
+                kid: recoveryKeyId
+              },
+              key
+            };
+            const encryptOptions = {
+              contentAlg: 'A256GCM',
+              format: 'compact'
+            };
+
+            if (options.iv) {
+              encryptOptions.iv = options.iv;
+            }
+
+            return jose.JWE.createEncrypt(encryptOptions, recipient)
+              .update(JSON.stringify(recoveryData))
+              .final();
+          });
+      });
+    },
+
+    /**
+     * Unbundle the encoded recovery data using the specified recovery key.
+     *
+     * @param {String} recoveryEnc Recovery encryption key
```

Comment:
Some comment as above about `recoveryEnc`, that name isn't clear to me since the description says this is a key. Can you rename the variable `recoveryEncKey`? Though this operation decrypts the data, so perhaps the variable should just be called `recoveryKey` universally?